src/BotPlutusInterface/CoinSelection.hs

```diff
@@ -0,0 +1,479 @@
+selectTxIns originalTxIns utxosIndex outValue =
+  runEitherT $ do
+    let -- This represents the input value.
+        txInsValue :: Value
+        txInsValue =
+          foldOf (folded . to ((`Map.lookup` utxosIndex) . txInRef) . folded . to txOutValue) originalTxIns
+
+        -- This is set of all the asset classes present in outValue, inputValue and all the utxos combined
+        allAssetClasses :: Set AssetClass
+        allAssetClasses =
+          uniqueAssetClasses $ txInsValue : outValue : utxosIndex ^.. folded . to txOutValue
+
+        txInRefs :: [TxOutRef]
+        txInRefs = originalTxIns ^.. folded . to txInRef
+
+        -- All the remainingUtxos that has not been used as an input to the transaction yet.
+        remainingUtxos :: [(TxOutRef, TxOut)]
+        remainingUtxos =
+          Map.toList $
+            Map.filterWithKey
+              (\k v -> k `notElem` txInRefs && isRight (txOutToTxIn (k, v)))
+              utxosIndex
+
+    lift $ printBpiLog @w (Debug [CoinSelectionLog]) $ "Remaining UTxOs: " <+> pretty remainingUtxos
+
+    -- the input vector for the current transaction, this can be a zero vector when there are no
+    -- inputs the transaction.
+    txInsVec <-
+      hoistEither $
+        if Value.isZero txInsValue
+          then Right $ zeroVec (length allAssetClasses)
+          else valueToVec allAssetClasses txInsValue
+
+    -- the output vector of the current transaction, this is all the values of TxOut combined.
+    outVec <- hoistEither $ valueToVec allAssetClasses outValue
+
+    -- all the remainingUtxos converted to the vectors.
+    remainingUtxosVec <- hoistEither $ mapM (valueToVec allAssetClasses . txOutValue . snd) remainingUtxos
+
+    -- we use the default search strategy to get indexes of optimal utxos, these indexes are for the
+    -- remainingUtxos, as we are sampling utxos from that set.
+    selectedUtxosIdxs <- newEitherT $ searchTxIns @w def (isSufficient outVec) outVec txInsVec remainingUtxosVec
+
+    lift $ printBpiLog @w (Debug [CoinSelectionLog]) $ "" <+> "Selected UTxOs Index: " <+> pretty selectedUtxosIdxs
+
+    let -- These are the selected utxos that we get using `selectedUtxosIdxs`.
+        selectedUtxos :: [(TxOutRef, TxOut)]
+        selectedUtxos = selectedUtxosIdxs ^.. folded . to (\idx -> remainingUtxos ^? ix idx) . folded
+
+        selectedVectors :: [Vector Integer]
+        selectedVectors = selectedUtxosIdxs ^.. folded . to (\idx -> remainingUtxosVec ^? ix idx) . folded
+
+    finalTxInputVector <- hoistEither $ foldM addVec txInsVec selectedVectors
+    unless (isSufficient outVec finalTxInputVector) $ throwError "Insufficient Funds"
+
+    selectedTxIns <- hoistEither $ mapM txOutToTxIn selectedUtxos
+
+    lift $ printBpiLog @w (Debug [CoinSelectionLog]) $ "Selected TxIns: " <+> pretty selectedTxIns
+
+    -- Now we add the selected utxos to originalTxIns present in the transaction previously.
+    return $ originalTxIns <> Set.fromList selectedTxIns
```

Comment:
Completely stylistic, but: for consistency with the rest of the codebase (and because `return` is a historical thing that's no longer necessary):
```suggestion
    pure $ originalTxIns <> Set.fromList selectedTxIns
```
(there are some other `return`s in this module but I won't comment on them all)

Reply:
fixed